### ISSUE TITLE
Support IncludeIndependantAssociations for EF Core

### ISSUE THREAD

**Issue by mrlife** [thepirat000/Audit.NET#381]
With EF Core 5 [now supporting](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-5.0/whatsnew#many-to-many) many-to-many relationships without explicitly mapping the join table, can this be revisited?

It may be possible to find the needed associations here:

```c#
// get entries
var entityEntries = ChangeTracker.Entries();

// get one entry
EntityEntry entityEntry = entityEntries.FirstOrDefault();

// get collections
var collectionEntries = entityEntry.Collections;

// get one collection
CollectionEntry collectionEntry = collectionEntries.FirstOrDefault();
```

[EntityEntry](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.changetracking.entityentry?view=efcore-5.0)

[CollectionEntry](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.changetracking.collectionentry?view=efcore-5.0)

**Comment by thepirat000** [comment]
For EF Core 5 the library already includes the associations as normal entries on the Audit Event.

For example, if you model a many to many relation like this:

```c#
public class Post
{
    public int Id { get; set; }
    public string Name { get; set; }
    public ICollection<Tag> Tags { get; set; }
}

public class Tag
{
    public int Id { get; set; }
    public string Text { get; set; }
    public ICollection<Post> Posts { get; set; }
}
```

And then you insert one Post related to one Tag, like this:

```c#
using (var context = new ManyToManyContext())
{
    var post = new Post()
    {
        Id = 10,
        Name = "test",
        Tags = new List<Tag>() { new Tag() { Id = 101, Text = "tag 1" } }
    };
    context.Posts.Add(post);
    context.SaveChanges();
}
```

You will obtain an EF audit event which has 3 entries:

- Insert to Post table
- Insert to Tag table
- Insert to the relation PostTag table 

The audit event output will look like this:

```javascript
"EntityFrameworkEvent": {
	"Database": "ManyToMany",
	"ConnectionId": "f78f6b97-4451-46b4-9c39-1d1faa109227",
	"Entries": [
		{
			"Table": "Posts",
			"Name": "Post",
			"Action": "Insert",
			"PrimaryKey": {
				"Id": 10
			},
			"ColumnValues": {
				"Id": 10,
				"Name": "test"
			},
			"Valid": true
		},
		{
			"Table": "Tags",
			"Name": "Tag",
			"Action": "Insert",
			"PrimaryKey": {
				"Id": 101
			},
			"ColumnValues": {
				"Id": 101,
				"Text": "tag 1"
			},
			"Valid": true
		},
		{
			"Table": "PostTag",
			"Name": "PostTag (Dictionary<string, object>)",
			"Action": "Insert",
			"PrimaryKey": {
				"PostsId": 10,
				"TagsId": 101
			},
			"ColumnValues": {
				"PostsId": 10,
				"TagsId": 101
			},
			"Valid": true
		}
	],
	"Result": 3,
	"Success": true
}
```

**Comment by mrlife** [comment]
Ok, I see. I think this confused me from the README. I thought that this wasn't supported yet.

> IncludeIndependantAssociations: Value to indicate if the Independant Associations should be included. Default is false. (Only for EF <=6.2, **since EF Core does not support many-to-many relations without a join entity**.).

So, `IncludeIndependantAssociations` isn't needed on EF Core 5, awesome.

To audit the automatically created join table for many-to-many relationships, e.g. `PostTag`, does Audit.NET need us to create the join entity and map it? Can this be done automatically since it is "mapped" by the entities in the relationship and the join entity isn't needed by EF Core to create the join table?

**Comment by thepirat000** [comment]
Yes. I will update the readme since it was written before ef core 5 release and it's confusing.

For the other question I think you're referring to the EF data provider having a 1-1 relation between audited entity and audit table, right?

**Comment by mrlife** [comment]
> For the other question I think you're referring to the EF data provider having a 1-1 relation between audited entity and audit table, right?

I _think_ we are talking about the same thing. Just trying to get the audit tables for the join table.

**Comment by thepirat000** [comment]
I don't think there is a way for the EF data provider to automatically map something that is unrelated to the model.

But one way to map entries with implicitly created join tables would be to provide a generic Audit Entity Creator for such entries that are **not** related to entity types. I mean the user to provide a function that receives the unmapped entry and constructs the audit entity that needs to be saved. 

For example using same model as in my previous answer and assuming the following Audit entities:

```c#
[AuditIgnore]
public class Audit_Post
{
    public int Id { get; set; }
    public string Action { get; set; } // Insert, Delete or Update
    public int PostId { get; set; }
    public string Name { get; set; }
}
[AuditIgnore]
public class Audit_Tag
{
    public int Id { get; set; }
    public string Action { get; set; }
    public int TagId { get; set; }
    public string Text { get; set; }
}
[AuditIgnore]
public class Audit_PostTag
{
    public int Id { get; set; }
    public string Action { get; set; }
    public int TagId { get; set; }
    public int PostId { get; set; }
}
```

So if the library provides a new "ExplicitAuditEntityCreator" configuration, you could do something like this:

```c#
Audit.Core.Configuration.Setup()
    .UseEntityFramework(_ => _
        .AuditTypeExplicitMapper(map => map
            .Map<Post, Audit_Post>((Post post, Audit_Post auditPost) =>
            {
                auditPost.PostId = post.Id;
                auditPost.Name = post.Name;
            })
            .Map<Tag, Audit_Tag>((Tag tag, Audit_Tag auditTag) =>
            {
                auditTag.TagId = tag.Id;
                auditTag.Text = tag.Text;
            })
            .ExplicitAuditEntityCreator((AuditEvent ev, EventEntry ent) =>
            {
                // Explicit audit entity creation, this will only execute for unmapped entities
                if (ent.Table == "PostTag")
                {
                    return new Audit_PostTag() { Action = ent.Action, PostId = (int)ent.ColumnValues["PostsId"], TagId = (int)ent.ColumnValues["TagsId"] };
                }
                return null;
            })
            .AuditEntityAction((ev, ent, obj) =>
            {
                ((dynamic)obj).Action = ent.Action;
            }))
        .IgnoreMatchedProperties(true));
```

So if I run the following:

```c#
using (var context = new MyContext())
{
    // Add Post 10 related to tag 101 and 102
    var tag1 = new ManyToManyContext.Tag() { Id = 101, Text = "tag 1" };
    var tag2 = new ManyToManyContext.Tag() { Id = 102, Text = "tag 2" };
    var post = new ManyToManyContext.Post()
    {
        Id = 10,
        Name = "test",
        Tags = new List<ManyToManyContext.Tag>() { tag1, tag2 }
    };
    context.Posts.Add(post);
    context.SaveChanges();

    // Remove Post 10 relation with tag 101
    post = context.Posts.Include(p => p.Tags).FirstOrDefault(p => p.Id == 10);
    post.Tags.Remove(post.Tags.First(t => t.Id == 101));
    context.SaveChanges();
}
```

Then I get the following on the audit tables:

![image](https://user-images.githubusercontent.com/9836380/111888526-fbf86780-89a2-11eb-94d8-0456d9677f9f.png)


Do you think this would work for your use case?

**Comment by mrlife** [comment]
@thepirat000 Thank you for this. I think opting in to `IgnoreMatchedProperties(true)` for all audit entities would add some extra complexity for a large audited model. Would any of these ideas be a good match for the library?

1. Ability to only ignore matched properties for the `ExplicitAuditEntityCreator()`, either always or with a boolean parameter to that method
2. A new `MapExplicit()` that takes string names for the models when there is no explicit join model/entity. `"PostTag"` would map to the `"Table": "PostTag"` of the audit even output from your previous answer (aka when `ent.Table == "PostTag"` from your most recent answer).

```c#
// passes postTag where postTag.Table == "PostTag"
.MapExplicit("PostTag", "Audit_PostTag", (postTag, auditPostTag) =>
{
    auditPostTag.PostId = postTag.PostsId;
    auditPostTag.TagId = postTag.TagsId;
})
```

**Comment by thepirat000** [comment]
Thanks for sharing your ideas.

For the first point, the property matching will not execute for the entities passed to `ExplicitAuditEntityCreator` (regardless the `IgnoreMatchedProperties` configuration value), since there is no source entity type, so there are no properties to match.

For the second point, since there is no `PostTag` class, I guess the only way to pass that "postTag" parameter to the `MapExplicit` would be to make it `dynamic` and prefill with the column names as properties, but losing compile-time type checking and probably creating new problems. 

What about something like this, passing the column names as a dictionary instead? 

```c#
.MapExplicit<Audit_PostTag>("PostTag", (Dictionary<string, object> columns, Audit_PostTag auditPostTag) =>
{
    auditPostTag.PostId = columns["PostsId"];
    auditPostTag.TagId = columns["postTag"];
})
```

**Comment by thepirat000** [comment]
I was making some tests and I think I will be able to provide the following functionality on the Explicit Mapper (`AuditTypeExplicitMapper`):

Map explicitly by table name to a target type:

```c#
.MapTable<Audit_PostTag>("PostTag")
```

Map by table name providing a specific action:

```c#
.MapTable("PostTag", (EventEntry ent, Audit_PostTag auditPostTag) =>
{
    auditPostTag.SomeProperty = ent.ColumnValues["column name"];
})
```

Or for more complex predicates:

```c#
.MapExplicit<Audit_PostTag_Inserts>(ent => ent.Action == "Insert" && ent.Table = "PostTag")
```

Note that if the Property Matching is enabled it will try to map the entry Column values to the target audit type Properties where the Column.Name == Property.Name

Also note the IgnoreMatchedProperties can be setup to decide by target audit type, for example:

```c#
.IgnoreMatchedProperties(t => !(t == typeof(Audit_PostTag))));
```

means the property matching is enabled only for the entries that maps to a `Audit_PostTag` object.

**Comment by mrlife** [comment]
@thepirat000 This is very impressive, yet again. The `MapTable<T>()` should work well for my scenario (implicitly created join table for many-to-many relationship). I will try it out as soon as it's available.

Along with the other additions of `MapExplicit<T>()` and `IgnoreMatchedProperties(t => ...)`, it seems the library can cover almost any scenario.

I hope it is not too much trouble. If I can help in some way, please let me know.

**Comment by thepirat000** [comment]
This was implemented on version 16.5.5, please upgrade your references and re-test using `MapExplicit()` or `MapTable()`

**Comment by mrlife** [comment]
I tested out using `MapTable()` and it works great! Thanks again.
